plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts

```diff
@@ -74,7 +73,7 @@ export class CookieCutter implements TemplaterBase {
         logStream,
       });
     } else {
-      await runDockerContainer({
+      await containerRunner.runContainer({
         imageName: imageName || 'spotify/backstage-cookiecutter',
         args: [
           'cookiecutter',
```

Comment:
@fox32 noted that this could maybe also take over the local command execution.

Instead of this:
```ts
const isInstalled = await commandExists('cookiecutter');
if (isInstalled) {
  await runCommand(…);
} else {
  await runContainer(…);
}
```

Use a more generic runner interface that could handle the check if something can be executed locally.
Something like this:
```ts
const commandRunner: CommandRunnerApi = new CommandRunnerMux({
  runners: [
    // use one or the other, or both with a fallback? Could be configured from the integrators backend.
    new CommandRunnerLocal(),
    new CommandRunnerDocker(),
  ]
});

// …

await commandRunner.runCommand({
  // how to run locally?
  localCommand: 'cookiecutter',

  // how to run in a container runtime?
  imageName: 'spotify/backstage-cookiecutter',
  imageCommand: 'cookiecutter',

  // generic args that apply to both
  args: […],
  
  // other props like input/output dir, …
}
```

But I'm not sure if we want to have this kind of generalization. I wonder if there is already a library that implements such a pattern that could just be reused.

Reply:
Hmm, I like the idea even more now you wrote it down 😆

Reply:
Then again, not so neat that the call to the supposedly generic interface needs to have knowledge of the underlying implementations :) (both local and image based ones)

Reply:
Correct. My hope would be that we can find a common baseline of properties that are kind-of generic for the execution. 

Or we go with an even more generic option like `commandRunner.runCommand({command: 'cookiecutter', args: [...], directories: […], …});` and the runner can then decide to run the `cookiecutter` command in the `spotify/backstage-cookiecutter` image on the Docker runtime where the configuration is done outside of the techdocs/scaffolder plugin.

Reply:
We can leave this as a future improvement right? Basically letting the runner transparently run locally if needed. Only thing I could se we'd need to decide right now is the naming, for example should the interface be called `CommandRunner` instead? I'd prefer to just stick with the container oriented naming and then treat local command execution as more of an exception that you can register on the implementation. Something like the implementation being "here's how you run containers, but if the command is 'cookiecutter' then you run this separate thing instead".

Reply:
@Rugvip Something like this?

```ts
// decides if a command can be run locally or not
class LocalFallbackContainerRunner implements ContainerRunner {
  constructor(
    private readonly localRunner: ContainerRunner,
    private readonly fallbackRunner: ContainerRunner,
  ) {}

  async runContainer({command, ...}: ...) {
    if (await commandExists(command)) {
      await this.localRunner.runCommand(...);
    else {
      await this.fallbackRunner.runCommand(...);
    }
}

// Runs a command locally
class LocalContainerRunner implements ContainerRunner {
  async runContainer({command, args}: ...) {
    await runCommand({command, args, ...});
  }
}

async somePseudoUsage() {
  const containerRunner = new LocalFallbackContainerRunner(
    new LocalContainerRunner(...),
    new DockerContainerRunner(...),
  );

  // ...

  // no need to handle the "local" case
  await containerRunner.runContainer({
    imageName: imageName || 'spotify/backstage-cookiecutter',
    command: 'cookiecutter',
    args: [
      '--no-input',
      '-o',
      '/output',
      '/input',
      '--verbose',
    ],
    // ...
  });
}
```

This should work and could be a good alternative while sticking to the `docker run` logic. We would setup the `LocalFallbackContainerRunner` as a default in the example app but the integrator should later select the one that fits best to his environment.

Another large issue is the mapping of `/input` and `/output` folders and the construction of the `mountDirs`. This wouldn't work when run locally and is also problematic if run in other container environments.

---

We are in the process of testing Argo Workflows as a "container runtime" with this interface. I got it running for the Scaffolder but also want to try it with TechDocs. Since it is quite unique I'm not sure if we will contribute it to the core code but could rather add it to the contrib folder.

Let's wait until we have a working implementation and can finally decide which kinds of inputs and outputs we need. I would really like to replace the "run local" logic with the new `ContainerRunner` interface. I'll rename the interfaces as discussed in another comment and add a few more changes. Since this is a breaking change for the consumers, I would prefer to only require changes in existing Backstage installations once. Let me know if this plan is ok for you. I can share more updates hopefully in the next few days.

Reply:
Yep, something of that sort, and could even be more specific of the type `if (command === 'cookiecutter') { return runLocalCookiecutter(...) } ...`. Possibly registering a map of override runners per command, or whatever else makes sense :p

Regarding the folder thing that's something that we could potentially want to deal with right now? For example we could do this:

```ts
  await containerRunner.runContainer({
    imageName: imageName || 'spotify/backstage-cookiecutter',
    command: 'cookiecutter',
    args: ({ mounts }) => [
      '--no-input',
      '-o',
      mounts['/output'],
      mounts['/input'],
      '--verbose',
    ],
    // ...
  });
```

Reply:
We could just use the host path in the container so it doesn't matter where the command is executed. Nobody forces us to use the hardcoded `/input` as a target mount point.

One thing we thought about is whether we should want to use some kind of "working directory" that is mounted into the container instead of letting the user specify different mount points. In some environments, it would be necessary to copy all inputs and outputs to a specific location prior/after the execution (could be shared volume or even a S3). It would be very hard to know upfront which mounts are input and which are output folders. With a working directory, we could just move the whole thing back and forth. The scaffolder already works this way, but the TechDocs could must be changed to get into such a directory layout.

Reply:
To conclude the discussion here: I decided to not work towards more variable mounts and in a discussion with @Rugvip we agreed that we don't want to not solve the "run local" use case with a generic `ContainerRunner`. Instead, it would be better to e.g. split the existing [`CookieCutter`](https://github.com/backstage/backstage/blob/2b5506a5d70fae1291300432d22dffaed0835b32/plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts) class into a `DockerCookieCutterRunner` and a `LocalCookieCutterRunner`. This could then be exchanged with the already existing mechanisms: https://github.com/backstage/backstage/blob/2b5506a5d70fae1291300432d22dffaed0835b32/packages/backend/src/plugins/scaffolder.ts#L37-L42

These interfaces can provide a much more tool-specific interface with e.g. a `templateDir` and an `outputDir` so a potential `XYZCookieCutterRunner` has a fixed semantic to work with.

Reply:
Totally agree - i think the rub here is that the interface was at the wrong level to start with, to conveniently cover both cases